hashtable/chaining/hashmap.go

```diff
@@ -0,0 +1,154 @@
+package chaining
+
+const (
+	// KeyNotFound is a constant indicates that
+	// a key cannot be found in the HashMap.
+	KeyNotFound      = 0
+	defaultCapacity  = 8
+	upperLoadFactor  = 0.75
+	lowerLoadFactor  = 0.2
+	growShrinkFactor = 2
+)
+
+type node struct {
+	key   int
+	value int
+	next  *node
+	prev  *node
+}
+
+// HashMap is a struct for storing key-value data.
+type HashMap struct {
+	items []*node
+	size  int
+}
+
+// NewHashMap creates a new HashMap instance.
+func NewHashMap() *HashMap {
+	return &HashMap{
+		items: make([]*node, defaultCapacity),
+	}
+}
+
+// Insert adds a new pair of key-value to the HashMap.
+// If the key already exists, it overwrites the key's value.
+func (h *HashMap) Insert(key int, value int) {
+	if h.getLoadFactor() >= upperLoadFactor {
+		h.grow()
+	}
+
+	newNode := &node{
+		key:   key,
+		value: value,
+	}
+
+	foundNode, index := h.searchNode(key)
+	if foundNode != nil {
+		// overwrite value of this key
+		foundNode.value = value
+		return
+	}
+
+	h.size++
+	if curNode := h.items[index]; curNode == nil {
+		// first node in the linked list
+		h.items[index] = newNode
+	} else {
+		// last node in the linked list
+		for curNode.next != nil {
+			curNode = curNode.next
+		}
+		curNode.next = newNode
+		newNode.prev = curNode
+	}
```

Comment:
Just a suggestion but if you use sentinel node and `searchNode` returns a suitable node, you won't have to search again in this case.